lib/Sanger/CGP/NgsQc/ValidateSampleMetaAndBamHeader.pm

```diff
@@ -0,0 +1,755 @@
+
+  if ($is_normals !~ m/^yes$|^y$|^no$|^n$/i) {
+    $bam_state .= $FAIL_CODE_010."; ";
+  }
+
+  if (! -e $align_file) {
+    warn "can not find the file: $align_file\n";
+    return $bam_state.$FAIL_CODE_020.";";
+  }
+
+  warn "--------\nstart to validate alignment file: $align_file\n--------\n";
+  my $sam = Bio::DB::HTS->new(-bam => $align_file);
+  my @lines = split "\n", $sam->header->text;
+  my %headers; # to handle duplicates!
+
+  my (@RG_lines, @SQ_lines, @PG_lines);
+  for(@lines) {
+    push @RG_lines, $_ if(index($_,'@RG') == 0);
+    push @SQ_lines, $_ if(index($_,'@SQ') == 0);
+    push @PG_lines, $_ if(index($_,'@PG') == 0);
+  }
+
+  if(scalar @RG_lines != 0) {
+    my %reported_error;
+    for (@RG_lines) {
+      my ($id) = "$_\t" =~ m/ID:([^\t]+)/;
+      if (defined $id) {
+        if (! exists $headers{'RG'}{$id} ) {
+          $headers{'RG'}{$id} = 0; # handle duplicates also = 1 when rg found in a read.
+          my ($pl) = "$_\t" =~ m/PL:([^\t]+)/;
+          if (! defined $pl && ! exists $reported_error{$FAIL_CODE_070}) {
+            warn "no PL for RG $id.\n";
+            $reported_error{$FAIL_CODE_070} = 1;
+            $bam_state .= $FAIL_CODE_070."; ";
+          } else {
+            if ($pl !~ m/^CAPILLARY$|^ILLUMINA$|^LS454$|^SOLID$|^HELICOS$|^IONTORRENT$|^ONT$|^PACBIO$/ && ! exists $reported_error{$FAIL_CODE_080}) {
+              warn "bad PL for RG $id.\n";
+              $reported_error{$FAIL_CODE_080} = 1;
+              $bam_state .= $FAIL_CODE_080."; ";
+            }
+          }
+          my ($lb) = "$_\t" =~ m/LB:([^\t]+)/;
+          if (! defined $lb && ! exists $reported_error{$FAIL_CODE_060}) {
+            warn "no LB for RG $id.\n";
+            $reported_error{$FAIL_CODE_060} = 1;
+            $bam_state .= $FAIL_CODE_060."; ";
+          }
+
+          # my ($cn) = "$_\t" =~ m/CN:([^\t]+)/;
+          # warn "no CN for RG $id.\n" if (! defined $cn);
+          # $bam_state .= $FAIL_CODE_050."; " if (! defined $cn);
+          my ($sm) = "$_\t" =~ m/SM:([^\t]+)/;
+          if (! defined $sm && ! exists $reported_error{$FAIL_CODE_090}) {
+            warn "no SM for RG $id.\n";
+            $reported_error{$FAIL_CODE_090} = 1;
+            $bam_state .= $FAIL_CODE_090."; ";
+          } elsif ($sm ne $sample_ID && ! exists $reported_error{$FAIL_CODE_100}) {
+            warn "bad SM for RG $id.\n";
+            $reported_error{$FAIL_CODE_100} = 1;
+            $bam_state .= $FAIL_CODE_100."; ";
+          }
+        } else {
+          if (! exists $reported_error{$FAIL_CODE_031}) {
+            warn "duplicated RG ID in header.\n";
+            $reported_error{$FAIL_CODE_031} = 1;
+            $bam_state .= $FAIL_CODE_031."; ";
+          }
+        }
+      }
+    }
+  } else {
+    warn "no RG lines!\n";
+    $bam_state .= $FAIL_CODE_030."; ";
+  }
+
+  my $enough_reads = bam_has_reads_more_than_threshold($align_file, $count_base_number);
+  if ($enough_reads ne 'yes') {
```

Comment:
This function seems to be returning text or a number depending on the result.  A state shouldn't really change the data type, suggest:

```perl
my ($enough_reads, $read_count) = bam_has_reads_more_than_threshold($align_file, $count_base_number);
if ($enough_reads == 0) {
  warn "Alignment file: $align_file has too few reads ($read_count reads found).\n";
  $bam_state .= $FAIL_CODE_021."; ";
}
```